File: 04_hamming_code/HammingCode.java
// CN Lab Experiment 4: Hamming Code Generation and Error Correction
// Uses even parity and corrects one-bit errors.

import java.util.Scanner;

public class HammingCode {
    static boolean isPowerOfTwo(int number) {
        return number != 0 && (number & (number - 1)) == 0;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter binary data: ");
        String data = scanner.nextLine().trim();

        int m = data.length(); // Number of data bits.
        int r = 0;             // Number of parity bits.

        while (Math.pow(2, r) < m + r + 1) { // Find enough parity bits.
            r++;
        }

        int n = m + r;
        char[] code = new char[n + 1]; // Use positions 1..n.
        for (int i = 0; i <= n; i++) code[i] = '0';

        int dataIndex = 0;
        for (int position = 1; position <= n; position++) { // Fill data bits.
            if (!isPowerOfTwo(position)) {
                code[position] = data.charAt(dataIndex);
                dataIndex++;
            }
        }

        for (int parity = 0; parity < r; parity++) { // Calculate parity bits.
            int p = 1 << parity;
            int parityValue = 0;

            for (int position = 1; position <= n; position++) { // Check covered bits.
                if ((position & p) != 0) {
                    parityValue ^= code[position] - '0';
                }
            }

            code[p] = (char) ('0' + parityValue);
        }

        StringBuilder encoded = new StringBuilder();
        for (int i = 1; i <= n; i++) encoded.append(code[i]);
        System.out.println("Hamming code: " + encoded);

        System.out.print("Enter received code (or press Enter to use generated code): ");
        String receivedInput = scanner.nextLine().trim();
        if (receivedInput.isEmpty()) receivedInput = encoded.toString();

        char[] received = new char[receivedInput.length() + 1];
        received[0] = '0';
        for (int i = 1; i < received.length; i++) {
            received[i] = receivedInput.charAt(i - 1);
        }

        int errorPosition = 0;

        for (int parity = 0; parity < r; parity++) { // Recalculate parity checks.
            int p = 1 << parity;
            int parityValue = 0;

            for (int position = 1; position < received.length; position++) {
                if ((position & p) != 0) {
                    parityValue ^= received[position] - '0';
                }
            }

            if (parityValue != 0) errorPosition += p; // Build error position.
        }

        if (errorPosition == 0) {
            System.out.println("No single-bit error detected.");
        } else {
            System.out.println("Error position: " + errorPosition);
            received[errorPosition] =
                received[errorPosition] == '0' ? '1' : '0'; // Flip the bad bit.

            StringBuilder corrected = new StringBuilder();
            for (int i = 1; i < received.length; i++) corrected.append(received[i]);
            System.out.println("Corrected code: " + corrected);
        }

        scanner.close();
    }
}

File: 04_hamming_code/hamming_code.py
# CN Lab Experiment 4: Hamming Code Generation and Error Correction
# This implementation uses even parity.

def is_power_of_two(number):
    return number != 0 and (number & (number - 1)) == 0

data = input("Enter binary data: ").strip()
m = len(data)                           # Number of data bits.
r = 0                                    # Number of parity bits.

while 2 ** r < m + r + 1:               # Find enough parity positions.
    r += 1

n = m + r
code = ["0"] * (n + 1)                  # Position 0 is unused.

data_index = 0
for position in range(1, n + 1):         # Fill non-parity positions.
    if not is_power_of_two(position):
        code[position] = data[data_index]
        data_index += 1

for parity in range(r):                  # Calculate each parity bit.
    p = 2 ** parity
    parity_value = 0
    for position in range(1, n + 1):     # Check positions covered by this parity bit.
        if position & p:
            parity_value ^= int(code[position])
    code[p] = str(parity_value)

encoded = "".join(code[1:])
print("Hamming code:", encoded)

received = input("Enter received code (or press Enter to use generated code): ").strip()
if not received:
    received = encoded

received = "0" + received                # Make positions 1-based.
error_position = 0

for parity in range(r):                  # Recalculate every parity check.
    p = 2 ** parity
    parity_value = 0
    for position in range(1, len(received)):  # Check positions covered by this parity.
        if position & p:
            parity_value ^= int(received[position])
    if parity_value:
        error_position += p              # Add this parity position to the error index.

if error_position == 0:
    print("No single-bit error detected.")
else:
    print("Error position:", error_position)
    corrected = list(received)
    corrected[error_position] = "1" if corrected[error_position] == "0" else "0"
    print("Corrected code:", "".join(corrected[1:]))
